compositor_pipeline/src/queue.rs

```diff
@@ -0,0 +1,97 @@
+mod internal_queue;
+
+use std::{
+    sync::{Arc, Mutex},
+    thread::{self, sleep},
+    time::{Duration, Instant},
+};
+
+use compositor_common::{Frame, Framerate, InputId};
+use compositor_render::input_frames::InputFrames;
+use crossbeam_channel::{tick, unbounded, Receiver, Sender};
+
+use self::internal_queue::{InternalQueue, QueueError};
+
+const DEFAULT_BUFFER_DURATION: Duration = Duration::from_millis(10);
+
+pub struct Queue {
+    internal_queue: Arc<Mutex<InternalQueue>>,
+    check_queue_channel: (Sender<()>, Receiver<()>),
+    output_framerate: Framerate,
+    buffer_duration: Duration,
+}
+
+impl Queue {
+    pub fn new(output_framerate: Framerate) -> Self {
+        Queue {
+            internal_queue: Arc::new(Mutex::new(InternalQueue::new(output_framerate))),
+            check_queue_channel: unbounded(),
+            output_framerate,
+            buffer_duration: DEFAULT_BUFFER_DURATION,
+        }
+    }
+
+    pub fn add_input(&self, input_id: InputId) {
+        let mut internal_queue = self.internal_queue.lock().unwrap();
+        internal_queue.add_input(input_id);
+    }
+
+    #[allow(dead_code)]
+    pub fn remove_input(&self, input_id: InputId) {
+        let mut internal_queue = self.internal_queue.lock().unwrap();
+        // TODO: gracefully remove input - wait until last enqueued frame PTS is smaller than output PTS
+        internal_queue.remove_input(input_id);
+    }
+
+    #[allow(dead_code)]
+    pub fn start(&self, sender: Sender<InputFrames>) {
+        let (check_queue_sender, check_queue_receiver) = self.check_queue_channel.clone();
+        let internal_queue = self.internal_queue.clone();
+        let tick_duration = self.output_framerate.get_interval_duration();
+        let buffer_duration = self.buffer_duration;
+
+        thread::spawn(move || {
+            // Wait for first frame
+            check_queue_receiver.recv().unwrap();
+            sleep(buffer_duration);
```

Comment:
why we need sleep?

Reply:
I put it there for the ticker to match enforced buffer pushes (when `start.elapsed() + buffer_duration > buffer_pts` is true, because some inputs haven't pushed frames yet), but putting it here probably is a bad idea. 

</br>

Do you think that it should be placed in `"ticker thread"` like this:
```rust
    fn start_ticker(
        tick_duration: Duration,
        check_queue_sender: Sender<()>,
        buffer_duration: Duration,
    ) {
        thread::spawn(move || {
            sleep(buffer_duration);
            let ticker = tick(tick_duration);
            loop {
                ticker.recv().unwrap();
                check_queue_sender.send(()).unwrap();
            }
        });
    }
```

or maybe we should ignore it, since those first ticks won't `"push"` frames out of the queue, because of `start.elapsed() + buffer_duration > buffer_pts` condition?
I think leaving it in the `"ticker thread"` is the more appropriate solution.